### ISSUE TITLE
Proof of 0 == 1

### ISSUE THREAD

**Issue by matthiasgoergens** [0xPolygonMiden/miden-vm#605]
When you execute the following Miden Assembly program, you end up with a stack of all 0s:

```
begin
  if.true
    push.1
  end
end
```

However this [repository](https://github.com/matthiasgoergens/miden-collision) has a proof to show that the execution ends with a 1 on the stack.

**Comment by bobbinth** [comment]
Very cool! Without having looked too deeply into this, my guess is that you constructed two Miden VM programs which do different things but hash to the same value. This is actually not too difficult to do right now as `JOIN` and `SPLIT` blocks are hashed in exactly the same way (see [here](https://github.com/0xPolygonMiden/miden-vm/blob/9c8af67a7768b920772b3a42d89947f7f639e089/core/src/program/blocks/join_block.rs#L10)).

The first program would correspond to the MASM code above and would look like:
```
SPLIT
    SPAN push.1 END
ELSE
    SPAN noop END
END
```
And the second program would be:
```
JOIN
    SPAN push.1 END
    SPAN noop END
END
```
As mentioned above, both of these would currently hash to the same value - but the second program would actually leave 1 at the top of the stack.

If this is what's happening - it's pretty impressive as you can't compile the second program from MASM - so, you had to construct the program MAST manually.

If something else is going on, let us know :)

**Comment by matthiasgoergens** [comment]
Yes, that's exactly what I did. Good sleuthing!

I would have even told you, if you had asked. No need to guess. (I only made the issue like this to be in the zero knowledge spirit, and to have some fun.)

In general, I have the impression that there's a few more instances lurking where the code assumes that an attacker will craft malicious MASM programs, instead of going deeper. But that's just a gut feeling.

Another thing: at the moment when we execute a split node we implicitly generate and execute `assert` and `not assert` (or so) in the true and false branches. As a matter of design, I would put those instructions explicitly inside the AST. Mostly so that people using the AST as a compiler target have more freedom. And because explicit is better than implicit. (As a compiler target, it would be even better of split would allow more than two sub nodes, but our compiler can build a binary tree by itself just fine.)

About fixing the hash problem I exploited here: doing different affine transformations for split vs join might work without making things otherwise more complicated or much slower? But I'm no expert in that.

**Comment by ekmett** [comment]
In theory changing the hash from Hash(a,b) to, say, Hash(3a+1,b) for one of those and Hash(5a+2,b) for the other using different numbers and offsets for each node type should be enough to eliminate these spurious hash collisions. As an arithmetizable hash function it should robustly send linearly adjusted inputs to unrelated results, without having to widen Rescue Prime to take another input.

The reason for using an affine transformation rather than just multiplying by a constant is we saw you were using 0 as one of the rescue prime inputs in one of the hash cases.

The cost of an extra addition and multiplication in the (extended) field per instruction shouldn't be too bad, relative to the full price of the rescue prime hashing chiplet!

**Comment by ekmett** [comment]
(Concretely, I'd suggest changing to different affine offsets for one of the Rescue Prime input for all different syntax tree node types, that way you trivially satisfy the condition that all trees should hash to different values.) That's what I've been doing in my local testing.

**Comment by bobbinth** [comment]
@matthiasgoergens - it was fun to guess! :) I agree, there are probably a couple of more instances of where we need to ensure that an adversarial prover can't do anything malicious.

Regarding `SPLIT` nodes. I actually thought of doing something like this at the beginning but eventually decided to make asserts implicit. I'll need to think through the implications of making them implicit a bit more.

**Comment by bobbinth** [comment]
@ekmett yes, I think this should work. We've also been thinking of using "domain separation" to avoid these collisions. For example, setting one of capacity registers to 1 for `SPLIT` nodes, and to keeping at 0 for `JOIN` nodes. So, need to think through pros and cons of these approaches.

Overall, we are planning to fix this after we migrate to the RPO hash function (#474).

**Comment by matthiasgoergens** [comment]
I was suggesting explicit asserts in the split node, so that we can translate something like a switch-case or an indirect call nicer.

Basically on the basis of: guess the right answer and choose the right code path, then assert that you guessed right.

**Comment by ekmett** [comment]
The downside of domain separation is that the current hash function only has a fixed number of slots and you're using them all, so you either hash twice (expensive in time) or you lose precision on your hash values and hence your collision resistance (possibly very expensive for application users!). that's what led me towards affine transformations.

**Comment by matthiasgoergens** [comment]
@ekmett If you hash twice for eg `split`, you still have to be careful that this is not equivalent to nesting two `join` nodes (or some other shenanigans like that)?

**Comment by ekmett** [comment]
I meant if you have a hash function that takes 8 values down to 4, you can either do an affine transformation of the input (or output, which in retrospect is better), or you can replace one of those 8 values with something else, destroying some information. I prefer the affine form as a mixer. I have a WIP branch on my fork for swapping out the hashing function this way. I just need to fully map out how all the test code works. (Still have a few test failures to fix, and need to bring the branch up to date, as I forked it back when we first started talking about this.)

**Comment by bobbinth** [comment]
@ekmett thank you for submitting #607! I do want to spend a bit more time evaluating pros and cons of different approaches before settling on the one that you've implemented. The following are in the consideration set right now:

1. Affine transformation of the inputs (what you've suggested above). This results in a few more multiplications, but more importantly - I want to make sure it solves the problem (I think it does).
2. Affine transformation of the outputs (what you implemented in the PR). Similar considerations to above.
3. Using a different part of the state as return digest. Currently, we use element 4, 5, 6, and 7 as return digest, but we could  also use elements 8, 9, 10, 11. This requires no additional multiplications - but does complicate AIR constraints.
4. Using one of capacity registers for domain separation. For example, we could set element 2 to `ONE` when hashing a JOIN block and keep it as `ZERO` when hashing a split block. This would be the simplest and most efficient approach (it does not change the size of the input), but it does complicate the interface somewhat and also reduce security by 1 bit, I believe.

**Comment by ekmett** [comment]
I'm not wedded to the particular fix, but I wanted to at least get _some_ fix thrown over the fence so that I knew at least internally our team could proceed without this hanging over us.

**Comment by bobbinth** [comment]
Yep - this will definitely be fixed in v0.4 release.

**Comment by bobbinth** [comment]
With #682 merged, this should be fixed. Would be great to confirm and then we can close this issue.
